### PR TITLE
chore(ci): replace flake8 with Ruff as a linter

### LINKED ISSUE #2529: Install and configure Ruff

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #2380 

### Changes

This PR introduces the incredible project [Ruff](https://beta.ruff.rs/docs/) as the replacement for `flake8` and `isort` as our linter. By adopting Ruff, we gain access to numerous new checks that greatly enhance code maintainability, standards, docstyle, docstring quality, and more, benefitting from over 500 rules provided by Ruff. Also, Ruff is EXTREMELY fast compared to flake8.

**1 - Performance**
Ruff is fast, Ruff is very fast!! :flushed:. I did a comparison test running `make lint` using flake8 and Ruff. Ruff runs in less than 1 second, while flake8 takes around 25s to run.

**FLAKE8**
```sh
❯ time make lint                                                                                                 
poetry run isort aws_lambda_powertools tests examples
poetry run black aws_lambda_powertools tests examples
All done! ✨ 🍰 ✨
307 files left unchanged.
poetry run flake8 aws_lambda_powertools tests examples
make lint  25.74s user 0.38s system 360% cpu 7.253 total
```
**RUFF**
```sh
❯ time make lint
poetry run black aws_lambda_powertools tests examples
All done! ✨ 🍰 ✨
307 files left unchanged.
poetry run ruff aws_lambda_powertools tests examples
make lint  0.89s user 0.12s system 100% cpu 1.000 total
```
We still gain performance when running locally on the `pre-commit` hook, but since pre-commit only runs on modified files, the difference isn't too big. While Ruff takes less than 1s to inspect 10 files, flake8 takes 3s, but ok, we still gain performance and code standards.

**2 - Black and isort**
I suggest keeping black and removing isort for now. Ruff covers most (if not all) cases that isort covers, but the [documentation suggests](https://beta.ruff.rs/docs/faq/#is-ruff-compatible-with-black) we integrate Ruff and black

**3 - pydocstyle (D)**
We are failing most new rules checked by Ruff - https://beta.ruff.rs/docs/rules/#pydocstyle-d. If we enable all rules in pydocstyle, we will have around 5k errors to fix 🥲. I think we can check the most important and necessary, activate and fix the files.

**4 - Fixing files for new checks**
To provide clear explanations for the additions and removals of features supported by Ruff, I will comment out specific lines in the modified files. Considering the extensive changes required to be compliant with the new checks, it might be good to address them through separate PRs for better organization and review.

### User experience

There is no change in the user experience.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -17,16 +17,9 @@ repos:
...
-        language: system
-        types: [python]
-  - repo: local
-    hooks:
-      - id: flake8
-        name: linting::flake8
-        entry: poetry run flake8
+      - id: ruff
+        name: linting-format::ruff
+        entry: poetry run ruff
```
Comment (leandrodamascena) [inline .pre-commit-config.yaml:22]: Ruff reads the `ruff.toml` file and uses it as a configuration. We don't need to add a new entry in `pyproject.toml` to configure Ruff.

File: ruff.toml [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# Enable rules.
+select = [
+    "A",
+    "C4",
+    "C90",
+    "COM",
+    #"D", # pydocstyle - not enabled temporarily
```
Comment (leandrodamascena) [inline ruff.toml:8]: Check https://github.com/aws-powertools/aws-lambda-powertools-python/pull/2495#issue-1762230933

File: ruff.toml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    "W"
+]
+
+# Ignore specific rules
+ignore = [
+    "W291", # https://beta.ruff.rs/docs/rules/trailing-whitespace/
+    "PLR0913", # https://beta.ruff.rs/docs/rules/too-many-arguments/
+    "PLR2004", #https://beta.ruff.rs/docs/rules/magic-value-comparison/
+    "PLW0603", #https://beta.ruff.rs/docs/rules/global-statement/
+    "COM812", # Trailing comma missing - disabled temporarily
```
Comment (leandrodamascena) [inline ruff.toml:38]: When we enable this check, we will potentially have 494 errors to fix. I suggest creating a specific PR for this.

File: ruff.toml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    "PLC1901", # Compare-to-empty-string - disabled temporarily
+    "ERA001", # Found commented-out code - disabled temporarily
```
Comment (leandrodamascena) [inline ruff.toml:40]: When we enable this check, we will potentially have 25 errors to fix. I suggest creating a specific PR for this.

File: ruff.toml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    "PLW", # Warning category - disabled temporarily
+    "PLR", # Refactoring category - disabled temporarily
+    "PLC", # Convention category - disabled temporarily
+    "PLE", # Error category - disabled temporarily
```
Comment (leandrodamascena) [inline ruff.toml:44]: When we enable these checks, we will potentially have 30 errors to fix. I suggest creating a specific PR for these.

File: ruff.toml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    "ISC", # flake8-implicit-str-concat - disabled temporarily
+    "E501", # Line too long - disabled temporarily
```
Comment (leandrodamascena) [inline ruff.toml:46]: The way Ruff checks Line too long is a little different than flake8 does. Ruff doesn't consider` #noqa:E501` inside the comments and so we have to fix these files by adding `#noqa:E501 `at the end of the comment.

 I suggest creating a specific PR for this.

Reference: https://github.com/astral-sh/ruff/issues/2849

File: ruff.toml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+# Exclude files and directories
+exclude = ["docs", ".eggs", "setup.py", "example", ".aws-sam", ".git", "dist", ".md", ".yaml", "example/samconfig.toml", ".txt", ".ini"]
+
+# Maximum line length
+line-length = 120
+
+
+#fix = true
+#fixable = ["I", "COM812", "W"]
```
Comment (leandrodamascena) [inline ruff.toml:58]: After fixing all previous errors, we can enable automatic fix for specific checks.

File: ruff.toml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+# See: https://github.com/astral-sh/ruff/issues/128
+typing-modules = ["aws_lambda_powertools.utilities.parser.types"]
```
Comment (leandrodamascena) [inline ruff.toml:61]: We're importing `Literal` in `aws_lambda_powertools.utilities.parser.types` and exporting it to other files, so Ruff considers this a non-native Python type and complains.
Adding this as a typing module solved the problem.

File: ruff.toml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+max-complexity = 15
+
+[pylint]
+# Maximum number of nested blocks
+max-branches = 15
+# Maximum number of if statements in a function
+max-statements = 70
+
+[isort]
+split-on-trailing-comma = true
```
Comment (leandrodamascena) [inline ruff.toml:74]: We may need to check other configurations here: https://beta.ruff.rs/docs/settings/#isort

**Review by leandrodamascena — COMMENTED** [review]
Hi friends! Before moving forward with this PR, we need to set things straight. I made a lot of comments that we need to address.

**Comment by heitorlessa** [comment]
GREAT One. IIRC Ruff also replaces Bandit for SAST - could you check whether it accepts an existing baseline? If it does, you can definitely remove `bandit` too... if it doesn't, we can investigate in another PR.

Agree on `pydoc`. There's another big chunk of work for API documentation to be integrated into our User Guide docs, and that requires some cleaning and additions ;) We can do progressively.

https://github.com/aws-powertools/powertools-lambda-python/blob/develop/Makefile#L78

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -76,7 +76,7 @@ You might find useful to run both the documentation website and the API referenc
...
 | **Docstring**         | We use a slight variation of Numpy convention with markdown to help generate more readable API references.                                                                                                                                                                  |
-| **Style guide**       | We use black as well as flake8 extensions to enforce beyond good practices [PEP8](https://pep8.org/). We use type annotations and enforce static type checking at CI (mypy).                                                                                                |
+| **Style guide**       | We use black as well as [Ruff](https://beta.ruff.rs/docs/) extensions to enforce beyond good practices [PEP8](https://pep8.org/). We use type annotations and enforce static type checking at CI (mypy).                                                                                                |
```
Comment (heitorlessa) [inline CONTRIBUTING.md]:
```suggestion
| **Style guide**       | We use black as well as [Ruff](https://beta.ruff.rs/docs/) to enforce beyond good practices [PEP8](https://pep8.org/). We use type annotations and enforce static type checking at CI (mypy).                                                                                                |
```

small catch ;)

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -106,7 +106,7 @@ Looking at the existing issues is a great way to find something to contribute on
 
 This project has adopted the [Amazon Open Source Code of Conduct](https://aws.github.io/code-of-conduct).
 For more information see the [Code of Conduct FAQ](https://aws.github.io/code-of-conduct-faq) or contact
-opensource-codeofconduct@amazon.com with any additional questions or comments.
+<opensource-codeofconduct@amazon.com> with any additional questions or comments.
```
Comment (heitorlessa) [inline CONTRIBUTING.md]: NICE catch

File: ruff.toml [hunk 12]
```diff
@@ -0,0 +1,76 @@
+# Enable rules.
+select = [
+    "A",
+    "C4",
+    "C90",
+    "COM",
```
Comment (heitorlessa) [inline ruff.toml]: could you add a comment with that this means and/or link to Ruff docs for more context?

File: ruff.toml (hunk 9 above)
Comment (heitorlessa) [reply ruff.toml:74]: for `isort` we should use the profile option so it doesn't conflict with black later on imports.

`profile = "black"`

https://beta.ruff.rs/docs/faq/#how-does-ruffs-import-sorting-compare-to-isort

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
forgot to ask as the diff is huge but you prolly are removing all `flake8-` deps now, right?

File: ruff.toml (hunk 12 above)
Comment (leandrodamascena) [reply ruff.toml]: Done

**Comment by leandrodamascena** [comment]
> forgot to ask as the diff is huge but you prolly are removing all `flake8-` deps now, right?

I almost forgot about that, @heitorlessa! Done.

If you don't have any consideration, I think we can merge this PR and I'll open the others to enable rules and fix files.

File: ruff.toml (hunk 9 above)
Comment (leandrodamascena) [reply ruff.toml:74]: Ruff tries to keep the same profile as isort when using `profile="black"`. I think we don't need this configuration.

> Ruff's import sorting is intended to be near-equivalent to isort's when using isort's profile = "black".

 Besides that, Ruff doesn't support profile in isort configuration, @heitorlessa

>Ruff does not yet support all of isort's configuration options, though it does support many of them. You can find the supported settings in the [API reference](https://beta.ruff.rs/docs/settings/#isort).

```sh
❯ poetry run ruff aws_lambda_powertools tests examples 
error: Failed to parse `/home/leandro/DEVEL-PYTHON/lambda-powertools-oficial/ruff.toml`: TOML parse error at line 75, column 1
   |
75 | profile = "black"
   | ^^^^^^^

```

Reference: https://beta.ruff.rs/docs/faq/#how-does-ruffs-import-sorting-compare-to-isort

**Review by rubenfonseca — APPROVED** [review]
WOOOOOOOOOW THANK YOU FOR TAKING TIME TO DO THIS!! Looks awesome!

**Pull request merged**